cirq/google/devices/known_devices.py

```diff
@@ -60,56 +64,77 @@ def create_device_proto_from_diagram(
         ascii_grid: str,
         gate_sets: Optional[Iterable[
             serializable_gate_set.SerializableGateSet]] = None,
-        durations_picos: Dict[str, int] = None,
+        durations_picos: Optional[Dict[str, int]] = None,
+        out: Optional[device_pb2.DeviceSpecification] = None,
 ) -> device_pb2.DeviceSpecification:
-    """
-    Parse ASCIIart device layout into DeviceSpecification proto containing
-    information about qubits and targets.  Note that this does not populate
-    the gate set information of the proto.  This function also assumes that
-    only adjacent qubits can be targets.
+    """Parse ASCIIart device layout into DeviceSpecification proto.
 
-    Args:
-            ascii_grid: An ASCII version of the grid, see _parse_device for
-                    details
-            gate_set: A serializable gate_set object that is used to define
-                    the translation between gate ids and cirq Gate objects
-            durations_picos: A dictionary with keys of gate ids and values of
-                    the duration of that gate in picoseconds
+    This function assumes that all pairs of adjacent qubits are valid targets
+    for two-qubit gates.
 
+    Args:
+        ascii_grid: ASCII version of the grid (see _parse_device for details).
+        gate_sets: Gate sets that define the translation between gate ids and
+            cirq Gate objects.
+        durations_picos: A map from gate ids to gate durations in picoseconds.
+        out: If given, populate this proto, otherwise create a new proto.
     """
-
     qubits, _ = _parse_device(ascii_grid)
-    spec = device_pb2.DeviceSpecification()
 
-    # Create valid qubit list
     qubit_set = frozenset(qubits)
-    spec.valid_qubits.extend([v2.qubit_to_proto_id(q) for q in qubits])
+    pairs: List[Tuple['cirq.Qid', 'cirq.Qid']] = []
+    for qubit in qubits:
+        for neighbor in sorted(qubit.neighbors()):
+            if neighbor > qubit and neighbor in qubit_set:
+                pairs.append((qubit, neighbor))
+
+    return create_device_proto_for_qubits(qubits, pairs, gate_sets,
+                                          durations_picos, out)
+
+
+def create_device_proto_for_qubits(
```

Comment:
Optional: Consider adding a unit test. The function is covered by tests for `create_device_proto_from_diagram` so this is optional.

Benefits of adding a dedicated unit test:
 * we won't have to add it if/when `create_device_proto_from_diagram` is removed or changed not to rely on `create_device_proto_for_qubits`,
 * you get to test the very behavior you're interested in - filtering for specified qubits.

Reply:
Added a test of an "irregular" grid.